Merge Features/17 into dev branch

SniffBakaSniff opened the pull request.

SniffBakaSniff committed:
Implemented Ticket Systems Core Logic

I've added the main parts of the ticket system, including database services, ticket creation and deletion, and ticket claiming. It's still a work in progress and needs some finishing touches to be fully polished.

SniffBakaSniff committed:
Updated ButtonIntercations

Added permission checks for claiming a ticket and implemented a confirmation system when closing a ticket.

SniffBakaSniff committed:
Updated Confirmation System.

Confirmation now uses Buttons instead of DSharpPlus.Interactivity.

SniffBakaSniff committed:
Update ButtonInteractionListener.cs

Made it ephemeral and the cancel button is useful now.

SniffBakaSniff committed:
Added Join/leave message and customizable messages

Its a bit messy and basic but does the trick and shouldnt be to hard to refine later.

SniffBakaSniff committed:
Refactoring

Refactored the file structure and split DatabaseService into seperate files.

SniffBakaSniff committed:
Quick Fix to Last commit

Fixed some minor issues caused by the refactoring i did.

SniffBakaSniff committed:
Refactoring

Changed the name from ArtcordAdminBot to ArtcordBot.

SniffBakaSniff committed:
Additions to Ticket System

added commands for adding/removing users or roles to a ticket.

SniffBakaSniff committed:
Add exception handling

Added try-catch blocks to critical sections of the code to handle potential exceptions.

SniffBakaSniff committed:
Renamed csproj and sln files.

Renamed the csproj and sln files to match the project name.

Ascyt commented on Commands/Moderation/Ban Command/BanCommand.cs:
```diff
@@ -55,35 +57,40 @@ await _databaseService.NewBanRecordAsync(
             reason ??= "No reason provided.";
 
-            await ctx.Guild.BanMemberAsync(targetUser, TimeSpan.FromHours((int)deleteTimeframe), reason);
+            //await ctx.Guild.BanMemberAsync(targetUser, TimeSpan.FromHours((int)deleteTimeframe), reason);
```
old code

Ascyt commented on Commands/Moderation/Ban Command/BanCommand.cs:
````diff
@@ -55,35 +57,40 @@ await _databaseService.NewBanRecordAsync(
                 },
                 Timestamp = DateTime.UtcNow
-            };
+            }.AddField("Reason", $"```{reason}```");
````
Put new line characters between the \`\`\` things

Ascyt commented on Commands/Moderation/Ban Command/BanCommand.cs:
````diff
@@ -55,35 +57,40 @@ await _databaseService.NewBanRecordAsync(
                 },
                 Timestamp = DateTime.UtcNow
-            };
+            }.AddField("Reason", $"```{reason}```");
````
Here too

Ascyt commented on Commands/General/Ticket Command/TicketAddCommand.cs:
Not this file specifically, but rename the "Ticket Command" folder to "TicketCommand" or something similar, as spaces within file names are bad practices and can lead to issues

Ascyt commented on Commands/Miscellaneous/Ping Command/PingCommand.cs:
Rename the "Ping Command" folder to "PingCommand"

Ascyt commented on Commands/Moderation/ModCommandGroup.cs:
```diff
@@ -0,0 +1,22 @@
+        {
+            _banService = banService;
+            _guildSettingsService = guildSettingsService;
+            _httpClient = new HttpClient();
```
Curious as to what you need HttpClient for

SniffBakaSniff replied:
I dont remember. It prolly was for like the ban commands "proof" where you could upload an image or smt

Ascyt commented on Database/Models/TicketRecords.cs:
```diff
@@ -0,0 +1,32 @@
+    public ulong TicketId { get; set; }
+    public ulong UserId { get; set; }
+    public TicketMessageType MessageType { get; set; }
+    public string Content { get; set; } = null!;
```
Pretty sure you can use `public required string` to not have to do the `null!` thing

Ascyt commented on Database/Models/TicketRecords.cs:
```diff
@@ -0,0 +1,32 @@
+    public TicketMessageType MessageType { get; set; }
+    public string Content { get; set; } = null!;
+    public DateTime Timestamp { get; set; } = DateTime.UtcNow;
+    public TicketRecords TicketRecords { get; set; } = null!;
```
Same here

Ascyt commented on Helpers/MessageHelpers.cs:
```diff
@@ -20,6 +20,13 @@ public static DiscordEmbed GenericErrorEmbed(string message, string title = "Err
                 .Build();
-    }
+
+        public static DiscordEmbed GenericUpdateEmbed(string title, string extra = null!, string color = "#00ffff") => new DiscordEmbedBuilder()
```
Why do you have `string extra = null!` in the argument? Just make the string nullable by putting a question mark after it (`string?`)

SniffBakaSniff replied:
I have no idea. You think ik what im doing? 😂

Ascyt commented on Listeners/ButtonInteractionListener.cs:
Maybe split it up into multiple methods, because it's very long

Ascyt commented on Listeners/JoinLeaveListener.cs:
```diff
@@ -0,0 +1,57 @@
+            Console.WriteLine(welcomeMessage);
+            ulong? welcomeChannelId = await _guildSettingsService.GetWelcomeChannelAsync(e.Guild.Id);
+
+            if (welcomeChannelId.HasValue && welcomeMessage != null)
```
`is not` instead of `!=` to prevent operator overloading to be an issue

Ascyt commented on Listeners/JoinLeaveListener.cs:
```diff
@@ -0,0 +1,57 @@
+            ulong? farewellChannelId = await _guildSettingsService.GetFarewellChannelAsync(e.Guild.Id);
+            ulong? welcomeChannelId = await _guildSettingsService.GetWelcomeChannelAsync(e.Guild.Id);
+            
+            if (farewellChannelId.HasValue && farewellMessage != null)
```
```suggestion
            if (farewellChannelId.HasValue && farewellMessage is not null)
```

Ascyt commented on Listeners/JoinLeaveListener.cs:
```diff
@@ -0,0 +1,57 @@
+                
+                if (farewellChannel is not null && farewellChannel.Type == DiscordChannelType.Text)
+                {
+                    await farewellChannel.SendMessageAsync(farewellMessage.Replace("{server}", e.Guild.Name).Replace("{user}", e.Member.Mention));
```
Long line, use linebreaks to make it easier to read
```suggestion
                    await farewellChannel.SendMessageAsync(farewellMessage
                        .Replace("{server}", e.Guild.Name)
                        .Replace("{user}", e.Member.Mention));
```
below as well

Ascyt commented on Listeners/TicketMessageLogger.cs:
```diff
@@ -0,0 +1,27 @@
+            var messageType = e.Author.Id == userId ? TicketMessageType.Moderator : TicketMessageType.Submitter;
+            if(ticketId.HasValue) 
+            {
+                Console.WriteLine(e.Message.Content);
```
Console.WriteLine?

SniffBakaSniff replied:
I think that was temporary for debugging

SniffBakaSniff replied:
It has been awhile

Ascyt commented on Services/CustomPrefixResolver.cs:
```diff
@@ -4,24 +4,24 @@
     public async ValueTask<int> ResolvePrefixAsync(CommandsExtension extension, DiscordMessage message)
     {
-        if (string.IsNullOrWhiteSpace(message.Content) || message.Channel == null)
+        if (string.IsNullOrWhiteSpace(message.Content) || message.Channel! == null!)
```
`message.Channel! == null!` why the exclamation points? Can't you just `message.Channel is null`

SniffBakaSniff replied:
Bro idk what im doing lmao

Ascyt commented on Services/Database/GuildSettingsService.cs:
Not a huge fan of the way you're logging the exceptions here, maybe there's a way to catch them at a higher level and just print the stack trace, a lot of code duplication here

Ascyt commented on Services/Database/MessageSettingsService.cs:
```diff
@@ -0,0 +1,96 @@
+{
+    public class MessageSettingsService : IMessageSettingsService
+    {
+        public async Task<string?> ManageMessageSettingAsync(ulong guildId, string messageType, string? newMessage = null)
```
Pretty sure you can put an enum instead of a string for `messageType`

Ascyt reviewed (changes requested):
Pretty great stuff!!! Mostly lgtm, but a few things I've noted still. Importantly yeah, don't put spaces in file names as this can lead to issues

SniffBakaSniff committed:
Fixed various issues for PR

made the changes you wanted and added a ExceptionHandler class to reduce the amount of repetitive code for try-catch statements.

Ascyt commented on Services/Database/BanService.cs:
Couldn't you just put the class (`BanRecords`) itself in the arguments, instead of having one for every thing?

Ascyt replied:
Also I figure it should be called `BanRecord` and not `BanRecords` since it's only a single record

Ascyt commented on Services/Database/IDatabaseService.cs:
Yeah the arguments should be the classes themselves, not a single argument for every property

Ascyt reviewed (changes requested):
Pretty good, not much I would change👍

SniffBakaSniff committed:
Changes for PR #24 and new command

Made the changes that @Ascyt mentioned in Pull Request #24 and implemented a new command "/mod viewbans".

Ascyt reviewed (approved):
I think it's fine to merge. As I said in Discord, it'd still be cool if you made a general function for the pagination, but for now it's fine

SniffBakaSniff committed:
Implemented a new PaginationService.

Implemented a new `PaginationService` to retrieve paginated lists from the database. Updated the `/mod viewbans` command to include user lookup, allowing retrieval of ban records related to a specified user.

SniffBakaSniff commented:
Fancy stuff

SniffBakaSniff commented:
Huh

SniffBakaSniff commented:
Testing webhook

SniffBakaSniff commented:
Testing webhooks

Ascyt commented on Commands/Moderation/BanCommand/ViewBansCommand.cs:
```diff
@@ -0,0 +1,136 @@
+            {
+                var paginatedBans = await _paginationService.GetPaginatedResults(dbContext.BanRecords.Where(p => p.GuildId == ctx.Guild!.Id && p.UserId == userId), 1, 5);
+
+                if (paginatedBans.Records == null || !paginatedBans.Records.Any())
```
`is null` because operator overloading things

Ascyt commented on Listeners/ButtonInteractionListener.cs:
```diff
@@ -0,0 +1,229 @@
+                {
+                    var originalEmbed = e.Message.Embeds.FirstOrDefault();
+                    int currentPage = 1, totalPages = 1;
+                    if (originalEmbed != null && originalEmbed.Footer != null && !string.IsNullOrEmpty(originalEmbed.Footer.Text))
```
`is not null`

Ascyt commented on Listeners/ButtonInteractionListener.cs:
```diff
@@ -0,0 +1,229 @@
+                }
+
+                default:
+                    break;
```
Throw an error or something instead, for debugging purposes

Ascyt commented on Services/Database/BanService.cs:
```diff
@@ -0,0 +1,37 @@
+                using (var dbContext = new BotDbContext())
+                {
+                    var banRecord = await dbContext.BanRecords.FirstOrDefaultAsync(b => b.GuildId == guildId && b.UserId == userId);
+                    if (banRecord != null)
```
`is not null`

Ascyt reviewed (changes requested):
Good stuff, once you address the few changes you can merge it yourself👍

SniffBakaSniff committed:
Final Changes for Pull Request #24
